### ISSUE TITLE
vcpkg tool ignores builtin-baseline when downloading cmake, thus busting binary cache

### ISSUE THREAD

**Issue by GordonSmith** [microsoft/vcpkg#29090]
**Describe the bug**
Having set builtin-baseline for a manifest file, checking out a newer version of the vcpkg repo may cause the binary cache to invalidate due a change in the download cmake version.

**Environment**
- OS: Linux
- Compiler: NA

**To Reproduce**
1. Checkout vcpkg repo @ 2022.11.24 tagged version
2. Create a manifest with a `"builtin-baseline": "6f7ffeb18f99796233b958aaaf14ec7bd4fb64b2"` which corresponds to the tag for 2022.11.24
3. Do a `vcpkg install` 
4. Observe that the abi_info references cmake 3.24.0
5. Checkout vcpkg repo @ 2023.01.09
6. Do another `vcpkg install` with the same manifest - this causes a complete rebuild 
7. Observe that the abi_info now references cmake 3.25.0
8. Retry with `vcpkg install --x-abi-tools-use-exact-versions` 
9. Observe that the abi_info still references cmake 3.25.0

**Expected behavior**
I would expect vcpkg to download and use the version of cmake specified in the abi_info file.

**Failure logs**

**Additional context**

**Comment by autoantwort** [comment]
That is the indented behavior. The baseline is only used for the versions of the ports. If you want the exact same tools you have to use the same commit.

**Comment by dg0yt** [comment]
> to use the same commit.

... which will throw you back on broken msys2 downloads for windows? (#28919)

**Comment by GordonSmith** [comment]
@autoantwort - This will need to be elaborated on, as I am struggling to see how it is useful then?

My real world example:
* Currently checked out 2022.11.24 with matching builtin-baseline "6f7ffeb18f99796233b958aaaf14ec7bd4fb64b2"
* AWS no longer compiles on OSX builds (due to newer compiler in GH actions) - there is now a fix in vcpkg master branch (happens to be in the CURL dependency).
* I checkout master and set the version for Curl >= 3.6.1#2  (it was 3.6.1#1)

My binary cache is now bust and one of my other ports is failing due to an obscure change in cmake...

**Comment by autoantwort** [comment]
> which will throw you back on broken msys2 downloads for windows?

Yeah than this is not an option :) 

> This will need to be elaborated on, as I am struggling to see how it is useful then?

It is useful if you want the same version of the ports independent of what commit you check out.

**Comment by autoantwort** [comment]
> one of my other ports is failing due to an obscure change in cmake...

If you use the latest vcpkg master commit you don't have this problem.

**Comment by GordonSmith** [comment]
> > one of my other ports is failing due to an obscure change in cmake 3.25.0...
> 
> If you use the latest vcpkg master commit you don't have this problem.

As stated previously, this is a real world example and yes I do have this exact issue with master branch from this morning...

So the "correct" solution to my situation is copy / pasting the Curl port from master to my overlays folder on the older checkout (which to be fair is what I have had to do in the past).

If this is intended behaviour I will close this issue - but what am I missing as I am struggling to see when this could be useful in the real world?

**Comment by autoantwort** [comment]
> As stated previously, this is a real world example and yes I do have this exact issue with master branch from this morning...

Ah ok, because the latest vcpkg master uses cmake 3.25.1. But yeah it does probably does not fix your issue. I thought you references the popular cmake 3.25.0 bug on windows with the gcc and msvc abi. 

> So the "correct" solution to my situation is copy / pasting the Curl port from master to my overlays folder on the older checkout (which to be fair is what I have had to do in the past).

No using a baseline and a constraint as you wrote. 

> If this is intended behaviour I will close this issue - but what am I missing as I am struggling to see when this could be useful in the real world?

If you want to keep the versions of a port. 

The idea of versioning is to get selected versions, not to keep the binary cache hash stable.

**Comment by GordonSmith** [comment]
> > If this is intended behaviour I will close this issue - but what am I missing as I am struggling to see when this could be useful in the real world?
> 
> If you want to keep the versions of a port.
This doesn't explain in what scenario this behaviour is actually useful?

> The idea of versioning is to get selected versions, not to keep the binary cache hash stable.
That is fine for the sources - but in this case the change of cmake version causes a different header to be resolved by one of my find_packages, and the resulting lib is now out of sync with the vcpkg variants and is causing a runtime crash (not a compile time error).

The tagged version was tested with a known version of cmake.
That version of cmake is provided by vcpkg.
The version of cmake was stored in the abi_info file.

It was my assumption that once I locked my manifest at a specific builtin-baseline I checkout any newer commit hash on the same branch without breaking anything.
So even before I started to play with the version field, my app was failing.

If I want to force a diff version of cmake, I can enable the use local binaries flag, I am just at a loss that this is considered good?  I am sure one appropriate real world example will help?

**Comment by autoantwort** [comment]
> That is fine for the sources - but in this case the change of cmake version causes a different header to be resolved by one of my find_packages, and the resulting lib is now out of sync with the vcpkg variants and is causing a runtime crash (not a compile time error).

So this happened to one of your custom ports I guess? 

>  I would expect vcpkg to download and use the version of cmake specified in the abi_info file.

The abi_info file is never read (only hashed). vcpkg uses the systems cmake or downloads cmake specified in vcpkgTools.xml and does a version check of the found cmake executable if `--x-abi-tools-use-exact-versions` is passed. If a port is build the used cmake version is written to the abi_info file (and some other stuff). This file is then hashed and used as a key for binary caching.

**Comment by GordonSmith** [comment]
> > That is fine for the sources - but in this case the change of cmake version causes a different header to be resolved by one of my find_packages, and the resulting lib is now out of sync with the vcpkg variants and is causing a runtime crash (not a compile time error).
> 
> So this happened to one of your custom ports I guess?

Its a port from a different registry yes - but I would assume its only a matter of time until the latest master cmake version will stop working with older baselines?  Or is master tested against all the previous tagged ports?

> > I would expect vcpkg to download and use the version of cmake specified in the abi_info file.
> 
> The abi_info file is never read (only hashed). vcpkg uses the systems cmake or downloads cmake specified in vcpkgTools.xml and does a version check of the found cmake executable if `--x-abi-tools-use-exact-versions` is passed. If a port is build the used cmake version is written to the abi_info file (and some other stuff). This file is then hashed and used as a key for binary caching.

That answered my next question about why `--x-abi-tools-use-exact-versions` didn't do what I expected!

## Summary

### Current Behaviour (by design)
_Adding a builtin-baseline to your manifest will ensure the default port versions will be as off that commit hash (unless overridden specifically by a version >= or an overlay).  BUT will use the tooling as of the checked out version of vcpkg repo as specified in scripts/vcpkgTools.xml._

### My Desired Behaviour 
_Adding a builtin-baseline to your manifest will ensure the default port versions will be as off that commit hash (unless overridden specifically by a version >= or an overlay) **Including** all tooling._

### Workarounds

1. Copy and paste newer port into local overrides folder.
2. Checkout latest master, then replace the scripts folder with the one that matches the baseline in the manifest.

###  Question
Are there any real world examples where the current behaviour is beneficial?

**Comment by autoantwort** [comment]
> but I would assume its only a matter of time until the latest master cmake version will stop working with older baselines? 

Afaik cmake is backwards compatible. 
> Or is master tested against all the previous tagged ports?

No

> Are there any real world examples where the current behaviour is beneficial?

Yes. If you are on windows and would use the tools from the baseline, nothing would work because the downloads are gone (mysys removes stuff after some time).   
If you would use the versioning feature to get a newer version of a port and the port relies on a newer version of the tool, the port would break. The other way around (if you build an older version of a port with a newer version of the tool) it should work almost always since the tools are backwards compatible. 

Your case seems to be an exception and since you are the first person complaining about it so it does not seems to be a problem in general. Maybe you should report your problem at cmake that they should introduce a policy so that they don't break existing code.

**Comment by GordonSmith** [comment]
Yes I did hit the msys issue and the downloads expiring is also a general concern.  

You were also correct that we are considering the binary caching as a future proof "feature" (which was a false assumption).  I believe we can address that by removing the cmake version from the ABI (without too much concern if it is assumed to be backward compatible anyway?)

Thanks for your time it is appreciated and sorry if I came across as too opinionated, but we have a dev team of about 30 developers with different flavours of OSX / Windows and Linux and our vcpkg build time is >1 hour per developer (we do use NuGet caching, but only for the GH Action runners, which rarely matches our dev machines).
